File: js/agents.js
// Agent A (Bio-RL): dopamine-gated, reward-prediction-error plasticity on KC->MBON synapses inside the connectome.
// Agent B (Deep-RL): frozen connectome, REINFORCE on a linear softmax readout of descending-neuron spikes.
import { rng } from './sim.js';
import { WAIT } from './game.js';

// kcMbonGain 1: in the Janelia/Google male CNS the output neurons respond to visual input with no boost (FlyWire needed 8).
export const BIO_DEFAULTS = {
  kcMbonGain: 1, noiseHz: 400, noiseKick: 1.2, theta: 0.5, cRef: 2, eta: 0.15, mMax: 4, dopaHz: 150, sMax: 12,
};

export class BioRL {
  constructor(net, opts = {}) {
    this.o = { ...BIO_DEFAULTS, ...opts };
    this.kind = 'bio';
    this.rand = rng((opts.seed ?? 1) * 31 + 7);
    net.setNoise(net.idx.mbon, this.o.noiseHz, this.o.noiseKick);
    this.plasticPre = net.plasticPre;
  }

  decide(net) {
    const rates = net.groupMbon.map(ids => ids.reduce((s, i) => s + net.count[i], 0) / ids.length);
    const best = Math.max(...rates);
    const winners = [0, 1, 2].filter(k => rates[k] === best);
    const action = best >= this.o.theta ? winners[Math.floor(this.rand() * winners.length)] : WAIT;
    this.rates = rates;
    this.action = action;
    this.V = action === WAIT ? 0 : Math.tanh(best / this.o.cRef);
    if (action !== WAIT) this.elig = this.plasticPre[action].map(p => net.trace[p]); // eligibility snapshot
    return action;
  }

  reward(net, r) {
    this.delta = r - this.V; // reward-prediction error
    const hz = this.o.blocked ? 0 : this.o.dopaHz;
    net.setDrive(net.idx.pam, hz * Math.max(this.delta, 0) / 2);
    net.setDrive(net.idx.ppl1, hz * Math.max(-this.delta, 0) / 2);
  }

  afterDopamine(net, learn) {
    net.setDrive(net.idx.pam, 0);
    net.setDrive(net.idx.ppl1, 0);
    const sum = (ids) => ids.reduce((s, i) => s + net.count[i], 0);
    this.D = sum(net.idx.pam) / (net.idx.pam.length * this.o.sMax) - sum(net.idx.ppl1) / (net.idx.ppl1.length * this.o.sMax);
    if (!learn || this.action === WAIT) return; // no press -> no eligibility -> no weight change
    const edges = net.plastic[this.action], m = net.m, { eta, mMax } = this.o;
    for (let n = 0; n < edges.length; n++) {
      const k = edges[n], nm = m[k] + eta * this.D * this.elig[n];
      m[k] = nm < 0 ? 0 : nm > mMax ? mMax : nm;
    }
  }
}

// lr picked from a seed-1 sweep {0.03, 0.01, 0.003, 0.001} on the male CNS: highest mean of songs 26 to 30 (see docs/lab-notebook.md)
export const DEEP_DEFAULTS = { lr: 0.03, baselineRate: 0.05 }; // network built with BIO_DEFAULTS.kcMbonGain for both agents

export class DeepRL {
  constructor(net, opts = {}) {
    this.o = { ...DEEP_DEFAULTS, ...opts };
    this.kind = 'deep';
    this.rand = rng((opts.seed ?? 1) * 31 + 11);
    this.F = net.idx.dn.length;
    this.W = [0, 1, 2, 3].map(() => new Float32Array(this.F));
    this.b = new Float32Array(4);
    this.baseline = 0;
    this.mu = new Float32Array(this.F);
    this.var = new Float32Array(this.F).fill(1);
    this.seen = 0;
  }

  decide(net) {
    // Standardize each input with running statistics. Raw spike counts from hundreds of active neurons swamp the
    // softmax before it can learn, even though a readout told the answer decodes the lane from them perfectly.
    const raw = Float32Array.from(net.idx.dn, i => Math.log1p(net.count[i]));
    const k = 1 / Math.min(++this.seen, 200);
    for (let j = 0; j < this.F; j++) {
      const d = raw[j] - this.mu[j];
      this.mu[j] += k * d;
      this.var[j] += k * (d * (raw[j] - this.mu[j]) - this.var[j]);
    }
    const f = this.f = raw.map((v, j) => (v - this.mu[j]) / Math.sqrt(this.var[j] + 1e-3));
    const logits = this.W.map((w, a) => w.reduce((s, x, j) => s + x * f[j], this.b[a]));
    const mx = Math.max(...logits), ex = logits.map(z => Math.exp(z - mx)), Z = ex.reduce((s, x) => s + x, 0);
    this.p = ex.map(x => x / Z);
    let u = this.rand(), a = 0;
    while (a < 3 && u > this.p[a]) u -= this.p[a++];
    return (this.action = a);
  }

  reward(net, r) { this.r = r; }

  afterDopamine(net, learn) {
    if (!learn) return;
    const adv = this.r - this.baseline;
    this.baseline += this.o.baselineRate * (this.r - this.baseline);
    for (let a = 0; a < 4; a++) {
      const gcoef = this.o.lr * adv * ((a === this.action ? 1 : 0) - this.p[a]);
      this.b[a] += gcoef;
      const w = this.W[a];
      for (let j = 0; j < this.F; j++) w[j] += gcoef * this.f[j];
    }
  }
}

File: flyhero/agents.py
"""The two learners. js/agents.js is the browser port.

Bio-RL: dopamine-gated, reward-prediction-error plasticity on KC->MBON synapses inside the connectome.
Deep-RL: frozen connectome, REINFORCE on a linear softmax readout of descending-neuron spikes.
"""
import math

import numpy as np

from .game import WAIT
from .sim import Rng

# kc_mbon_gain 1: in the Janelia/Google male CNS the output neurons respond to visual input with no boost (FlyWire needed 8).
BIO_DEFAULTS = dict(kc_mbon_gain=1, noise_hz=400, noise_kick=1.2, theta=0.5, c_ref=2, eta=0.15, m_max=4, dopa_hz=150, s_max=12)


class BioRL:
    kind = 'bio'

    def __init__(self, net, seed=1, blocked=False, **opts):
        self.o = {**BIO_DEFAULTS, **opts}
        self.blocked = blocked
        self.rand = Rng(seed * 31 + 7)
        net.set_noise(net.idx['mbon'], self.o['noise_hz'], self.o['noise_kick'])
        self.action, self.V, self.delta = WAIT, 0.0, 0.0

    def decide(self, net):
        rates = [int(net.count[ids].sum()) / len(ids) for ids in net.group_mbon]
        best = max(rates)
        winners = [k for k in range(3) if rates[k] == best]
        action = winners[int(self.rand() * len(winners))] if best >= self.o['theta'] else WAIT
        self.rates, self.action = rates, action
        self.V = 0.0 if action == WAIT else math.tanh(best / self.o['c_ref'])
        if action != WAIT:
            self.elig = net.trace[net.plastic_pre[action]]  # eligibility snapshot (fancy indexing copies)
        return action

    def reward(self, net, r):
        self.delta = r - self.V  # reward-prediction error
        hz = 0 if self.blocked else self.o['dopa_hz']
        net.set_drive(net.idx['pam'], hz * max(self.delta, 0) / 2)
        net.set_drive(net.idx['ppl1'], hz * max(-self.delta, 0) / 2)

    def after_dopamine(self, net, learn):
        pam, ppl1, s_max = net.idx['pam'], net.idx['ppl1'], self.o['s_max']
        net.set_drive(pam, 0)
        net.set_drive(ppl1, 0)
        self.D = int(net.count[pam].sum()) / (len(pam) * s_max) - int(net.count[ppl1].sum()) / (len(ppl1) * s_max)
        if not learn or self.action == WAIT:
            return  # no press -> no eligibility -> no weight change
        k = net.plastic[self.action]
        net.m[k] = np.clip(net.m[k] + self.o['eta'] * self.D * self.elig, 0, self.o['m_max'])


# lr picked from a seed-1 sweep {0.03, 0.01, 0.003, 0.001} on the male CNS: highest mean of songs 26 to 30 (see docs/lab-notebook.md)
DEEP_DEFAULTS = dict(lr=0.03, baseline_rate=0.05)  # network built with BIO_DEFAULTS['kc_mbon_gain'] for both agents


class DeepRL:
    kind = 'deep'

    def __init__(self, net, seed=1, **opts):
        self.o = {**DEEP_DEFAULTS, **opts}
        self.rand = Rng(seed * 31 + 11)
        F = len(net.idx['dn'])
        self.W, self.b, self.baseline = np.zeros((4, F)), np.zeros(4), 0.0
        self.mu, self.var, self.seen = np.zeros(F), np.ones(F), 0

    def decide(self, net):
        # Standardize each input with running statistics. Raw spike counts from hundreds of active neurons swamp the
        # softmax before it can learn, even though a readout told the answer decodes the lane from them perfectly.
        raw = np.log1p(net.count[net.idx['dn']].astype(np.float64))
        self.seen += 1
        k = 1 / min(self.seen, 200)
        d = raw - self.mu
        self.mu += k * d
        self.var += k * (d * (raw - self.mu) - self.var)
        f = self.f = (raw - self.mu) / np.sqrt(self.var + 1e-3)
        # summed left to right like the JS; np.dot adds in a different order and rounds differently
        logits = [float(np.cumsum(np.concatenate(([self.b[a]], self.W[a] * f)))[-1]) for a in range(4)]
        mx = max(logits)
        ex = [math.exp(z - mx) for z in logits]
        Z = 0.0
        for x in ex:
            Z += x
        self.p = [x / Z for x in ex]
        u, a = self.rand(), 0
        while a < 3 and u > self.p[a]:
            u -= self.p[a]
            a += 1
        self.action = a
        return a

    def reward(self, net, r):
        self.r = r

    def after_dopamine(self, net, learn):
        if not learn:
            return
        adv = self.r - self.baseline
        self.baseline += self.o['baseline_rate'] * (self.r - self.baseline)
        for a in range(4):
            gcoef = self.o['lr'] * adv * ((1 if a == self.action else 0) - self.p[a])
            self.b[a] += gcoef
            self.W[a] += gcoef * self.f
